### PR TITLE
Select2 filter fixes

### PR DESCRIPTION
**Pull request by promatik**
This PR includes #3371 from @martijnb92. Fixing the dead loop when clearing filters.

This also fixes a behavior where `select2_ajax` filter leaves artifacts behind;

```
?select2_from_ajax=1&select2_from_ajax_text=Text.
```

When using the "Remove all filters" button, everything goes as expected. But when removing the entry directly on the filter it leaves behind `select2_from_ajax_text=Text.`

### CODE REVIEW

File: src/resources/views/crud/filters/select2_multiple.blade.php
Comment (promatik) [inline src/resources/views/crud/filters/select2_multiple.blade.php:91]:
```suggestion
	                    var values = $(this).val().filter(function(e){ return e === 0 || e });
```

File: src/resources/views/crud/filters/select2_ajax.blade.php
Comment (promatik) [inline src/resources/views/crud/filters/select2_ajax.blade.php:118]: By removing the validation to check if `val_text` exists, this function will run always, so when the value is removed it will clear the `filter_text` from URL.

**Comment by martijnb92** [comment]
Looks good!

**Comment by tabacitu** [comment]
Haha - chased quite a few PRs to get to this final solution. Good work @martijnb92 & @promatik ! Felt like a mouse in a maze - happy to see you have a solution you're happy with. Looks clean too!

I'll wait for @pxpm 's test, review & green light to merge this.

Cheers!

**Comment by pxpm** [comment]
@promatik 

I think we should merge this AFTER #3348 as there are some conflicts between the files (at first look line 112 will conflict with other PR)

What do you think, #3348 first then update this PR base branch to have the fixed code in #3348 or should we merge this one first and update the code in the other ? 

Or even merge this one in the other?

Let me know,
Pedro

**Comment by promatik** [comment]
@pxpm I think it's ok to wait for #3348 and fix the conflicts after that.

#3348 is _"Ready to Merge"_ I'll solve the conflicts and request your review after it's being merged 👌

**Comment by pxpm** [comment]
Hello @promatik 

I'v been working on this for the past days, we have also another issue open with filters #2772  that I think we should address in this PR too, because you know .. change files once, fix most possible stuff. 

1 issue raised by @AbbyJanke there was about `setModel` don't changing when applying the filter. 

From my tests:

```php
$this->crud->addFilter([ // dropdown filter
            'name' => 'select_from_array',
            'type' => 'dropdown',
            'label'=> 'Dropdown',
        ], ['one' => 'One', 'two' => 'Two', 'three' => 'Three'], function ($value) {
            if($value == 'one') {
                $this->crud->setModel('App\Models\Product');
            } else {
                $this->crud->setModel('App\Models\Icon');
            }
        });
```

When the filter is applied (the error is expected, because we don't have a "checkbox" column in Products), we are applying a filter, so the same filter logic is applied, but we can see that the `setModel` worked, because we are in monsters, querying for Products 👍 

image

The second issue was found while trying to debug the above mentioned. Backpack was doing more requests than it should when clearing the filters. 

You can easily reproduce it in https://demo.backpackforlaravel.com

If you go to monsters, apply any filter and then click `Remove Filters`, you will endup with this:

image

Problem relies in the fact that when clicking the button, we are going to "clear" all the filter values, and that produce a `change` event in some filters, that change event will eventually make the ajax request. Plus one more ajax request that we trigger (well done us) after clearing the filters.

We found a quick and dirt solution we posted on #2772 that was checking if there is value and returning if not. But that lead to undesired consequences because we also trigger a change event when you remove a single filter, and that would stop work.

So the solution is to when "Removing all filters", we update the browser url, removing the filters etc, but don't update the table url. The table will be updated in the triggered request after clearing all filters. 

Let me know @promatik @martijnb92 what you think, or any suggestions you have to improve this.

Best,
Pedro

**Comment by promatik** [comment]
Hi @pxpm I was reviewing your last changes, and it's great 👌 it clean up the code a lot.

I tested and found a bug with `shouldUpdateUrl`, it was not being reset so it worded only the first time.
Also I found a code to `e.stopPropagation();` on `select2_ajax`, it was causing the default `select2` to not clean the label when user click the clear button. I removed it and it caused no harm.
I fixed it, and along with that I refactored a little bit the function `updateDatatablesOnFilterChange` in a way that it makes it cleaner to call it.

So this;
```js
if(value === '' || !value) {
    var new_url = updateDatatablesOnFilterChange(crud, parameter, null, null, null, '{{ $filter->key }}', dateFilterShouldUpdateUrl)
} else {
    var new_url = updateDatatablesOnFilterChange(crud, parameter, value, null, null, '{{ $filter->key }}', true)
}
```

becomes;
```js
var new_url = updateDatatablesOnFilterChange(parameter, value, value || shouldUpdateUrl);
```

and in case of `select2_ajax`, call it twice;
```js
updateDatatablesOnFilterChange(filterName, val);
var new_url = updateDatatablesOnFilterChange(extra_param, val_text, true);
```
It's not amazing in this case, but it causes no harm and since it's the only situation where it's used it makes everything much cleaner.

I tested, everything is working, please let me know what you think about the changes 🙌

https://user-images.githubusercontent.com/1838187/108691461-72648f80-74f3-11eb-8444-75fed182d55e.mp4

File: src/resources/views/crud/filters/select2_ajax.blade.php
Comment (pxpm) [inline src/resources/views/crud/filters/select2_ajax.blade.php:145]:
```suggestion

```

**Review by pxpm — APPROVED** [review]

**Comment by tabacitu** [comment]
Great work here guys! I have only one concern... shouldn't we consider this a `breaking change` because `updateDatatablesOnFilterChange()` is provided in a separate file than a filter? 

Imagine this scenario:
- dev publishes and customizes `crud/inc/filters_navbar.blade.php` (for styling)
- we push this update
- from his point of view, all filters stop working

Don't get me wrong, I LIKE the fact that you've separated that repeating logic into `updateDatatablesOnFilterChange()`. But... I think we can only merge this in 4.2, due to that fact.

How big/common is this dead loop? Is it something that can wait until 4.2 to be merged? Or should we find a way to rewrite this in a non-breaking way?

**Comment by pxpm** [comment]
Hello @tabacitu 

You are right. Never came to my mind people overriding the `filters_navbar`.

I think the only breaking change here is the last fixes introduced in this PR that fix the multiple requests. The first issue fixed here was removing the `select2_ajax` filter parameters that got stuck in the URL. 

I will split this back into 2 PR's, one that could be merged in 4.1, and the refactor to be merged in 4.2. 

Thanks @tabacitu !

**Comment by pxpm** [comment]
Separated the non-breaking part of this PR into a new one. #3582 

Move this to 4.2.

**Comment by tabacitu** [comment]
@pxpm , it's now or never. If we don't merge this now, we'll separate some of these filters to `backpack/pro` and we'll never be able to do a clean merge.

Can you please:
- pull 4.2 on top of this, and fix merge conflicts;
- thoroughly test this - if it works, give me the green light to merge; if it does NOT work at first try, please don't spend hours one this; reply here that it no longer works and we'll figure out a solution after the 4.2 launch;

Thank you!

**Pull request merged**
